File: src/findsymbols.cc
/*
Copyright (C) 2003 Willem J. Atsma

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/
#include <octave/oct.h>

#ifdef NEED_OCTAVE_QUIT
#define OCTAVE_QUIT do {} while (0)
#else
#include <octave/quit.h>
#endif

#include <ginac/ginac.h>
#include "ov-vpa.h"
#include "ov-ex.h"
#include "symbols.h" 


/* Travel down the hierarchical expression and insert new symbols in the
   list so the endresult is sorted. */
static void append_symbols(octave_value_list& symlist,const GiNaC::ex& expression)
{
	int i, j, n = expression.nops();
	
	for(i=0;i<n;i++) {
		if(GiNaC::is_exactly_a<GiNaC::symbol>(expression.op(i))) {
			bool unique = true;
			int insert_here = symlist.length();
			GiNaC::ex ex_sym;
			GiNaC::symbol sym, sym_new = GiNaC::ex_to<GiNaC::symbol>(expression.op(i));
			std::string sym_name,sym_name_new = sym_new.get_name();
			for(j=0;j<symlist.length();j++) {
				OCTAVE_QUIT;
				/* have to convert back to compare: */
				get_symbol(symlist(j),ex_sym);
				sym = GiNaC::ex_to<GiNaC::symbol>(ex_sym);
				if(GiNaC::operator == (sym,sym_new)) {
					unique = false;
					break;
				} else {
					if(sym.get_name()>sym_name_new) {
						insert_here = j;
						break;
					}
				}
			}
			if(unique) {
				octave_value_list tmp = symlist;
				symlist.resize(symlist.length()+1);
				symlist(insert_here) = octave_value(new octave_ex(expression.op(i)));
				for(j=insert_here;j<tmp.length();j++)
					symlist(j+1) = tmp(j);
			}
		} else append_symbols(symlist,expression.op(i));
	}
}


DEFUN_DLD(findsymbols,args, ,
"-*- texinfo -*-\n\
@deftypefn {Loadable Function} {@var{syms} =} findsymbols(@var{f})\n\
\n\
Returns the symbols in symbolic expression @var{f} in a list.\n\
The list is sorted in alphabetical order.\n\
@end deftypefn")
{
	GiNaC::ex expression;
	octave_value retval;
	octave_value_list symlist;
	int nargin = args.length();

	if (nargin != 1) {
		error("Need one argument.");
		return retval;
	}
	try {
		if (!get_expression (args(0), expression)) {
			error("Argument must be a symbolic expression.");
			return retval;
		}
		/* Add 1 to so this works for symbols too. */
		append_symbols(symlist,expression+1);
		retval = symlist;
	} catch (std::exception &e) {
		error (e.what ());
		retval = octave_value ();
	}

	return retval;
}

File: @sym/findsymbols.m
%% Copyright (C) 2014 Colin B. Macdonald
%%
%% This file is part of Octave-Symbolic-SymPy
%%
%% Octave-Symbolic-SymPy is free software; you can redistribute
%% it and/or modify it under the terms of the GNU General Public
%% License as published by the Free Software Foundation;
%% either version 3 of the License, or (at your option) any
%% later version.
%%
%% This software is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty
%% of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
%% the GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public
%% License along with this software; see the file COPYING.
%% If not, see <http://www.gnu.org/licenses/>.

%% -- Loadable Function: L = findsymbols (X)
%%     Return a list (cell array) of the symbols in a symbolic expression
%%
%%     The list is sorted alphabetically..
%%
%%     Note E, I, pi, etc are not counted as symbols.

%% Author: Colin B. Macdonald
%% Keywords: symbolic, symbols, find

function L = findsymbols(x)


  % l.sort() fails on sympy 0.7.5
  cmd = [ 'def fcn(_ins):\n'                            ...
          '    x = _ins[0]\n'                           ...
          '    if not x.is_Matrix:\n'                   ...
          '        s = x.free_symbols\n'                ...
          '    else:\n'                                 ...
          '        s = set()\n'                         ...
          '        for i in x.values():\n'              ...
          '            s = s.union(i.free_symbols)\n'   ...
          '    l = list(s)\n'                           ...
          '    #dbout("warning: will fail on sympy 0.7.5")\n' ...
          '    #dbout(("we are", sp.__version__))\n' ...
          '    # FIXME: workaround for SymPy bug: #7413\n' ...
          '    l = sorted(l, key=str)\n'                ...
          '    return (l,)\n' ];

  L = python_sympy_cmd (cmd, x);


%!test
%! syms x b y n a arlo
%! z = a*x + b*pi*sin (n) + exp (y) + exp (sym (1)) + arlo;
%! s = findsymbols (z);
%! assert (isequal ([s{:}], [a,arlo,b,n,x,y]))
%!test
%! syms x
%! s = findsymbols (x);
%! assert (isequal (s{1}, x))
%!test
%! syms z x y a
%! s = findsymbols ([x y; 1 a]);
%! assert (isequal ([s{:}], [a x y]))
%!assert (isempty (findsymbols (sym (1))))
%!assert (isempty (findsymbols (sym ([1 2]))))
%!assert (isempty (findsymbols (sym (nan))))
%!assert (isempty (findsymbols (sym (inf))))
%!assert (isempty (findsymbols (exp (sym (2)))))

